rcl/src/rcl/arguments.c

```diff
@@ -1800,25 +1805,31 @@ _rcl_parse_remap_rule(
   output_rule->impl->replacement = NULL;
 
   rcl_lexer_lookahead2_t lex_lookahead = rcl_get_zero_initialized_lexer_lookahead2();
+  rcl_ret_t ret = rcl_lexer_lookahead2_init(&lex_lookahead, arg, allocator);
 
-  ret = rcl_lexer_lookahead2_init(&lex_lookahead, arg, allocator);
-  if (RCL_RET_OK != ret) {
-    return ret;
-  }
-
-  ret = _rcl_parse_remap_begin_remap_rule(&lex_lookahead, output_rule);
+  if (RCL_RET_OK == ret) {
+    ret = _rcl_parse_remap_begin_remap_rule(&lex_lookahead, output_rule);
 
-  if (RCL_RET_OK != ret) {
-    // cleanup stuff, but return the original error code
-    if (RCL_RET_OK != rcl_remap_fini(output_rule)) {
-      RCUTILS_LOG_ERROR_NAMED(ROS_PACKAGE_NAME, "Failed to fini remap rule after error occurred");
-    }
-    if (RCL_RET_OK != rcl_lexer_lookahead2_fini(&lex_lookahead)) {
-      RCUTILS_LOG_ERROR_NAMED(ROS_PACKAGE_NAME, "Failed to fini lookahead2 after error occurred");
+    rcl_ret_t fini_ret = rcl_lexer_lookahead2_fini(&lex_lookahead);
+    if (RCL_RET_OK != ret) {
+      if (RCL_RET_OK != fini_ret) {
+        RCUTILS_LOG_ERROR_NAMED(
+          ROS_PACKAGE_NAME, "Failed to fini lookahead2 after error occurred");
+      }
+    } else {
+      if (RCL_RET_OK == fini_ret) {
+        return RCL_RET_OK;
+      }
+      ret = fini_ret;
     }
-  } else {
-    ret = rcl_lexer_lookahead2_fini(&lex_lookahead);
   }
+
+  // cleanup output rule but keep first error return code
+  if (RCL_RET_OK != rcl_remap_fini(output_rule)) {
+    RCUTILS_LOG_ERROR_NAMED(
+      ROS_PACKAGE_NAME, "Failed to fini remap rule after error occurred");
+  }
```

Comment:
I'm not sure I agree with this logic, or, at least, I don't understand.

It looks to me with this change that we always cleanup output_rule, but that seems weird; we don't `remap_init` it in this method.  Shouldn't that be the responsibility of the caller?

In other words, it kind of looks to me like this method should be:

```
init_lexer();
if (failed_to_init_lexer)
  return fail;
parse_remap_begin_remap_rule(lexer, output_rule);
cleanup_lexer();
if (!ret) {
  if (!fini_ret)
    print_error();
  return ret;
}
```

And then the caller is responsible for `remap_fini` as appropriate.  Could you explain a little more what the idea is here?

Reply:
> It looks to me with this change that we always cleanup output_rule, but that seems weird

Only when something fails. 

> we don't remap_init it in this method.

We do, [right above](https://github.com/ros2/rcl/blob/6040298b429784ca9eb4d0404543d6211fd5f81d/rcl/src/rcl/arguments.c#L1792-L1800). Thing is, `remap_init` doesn't exist.

> Shouldn't that be the responsibility of the caller?

Right now it's not necessary. `_rcl_remap_parse_rule` takes a zero-initialized `rcl_remap_t` instance. On failure, it should stay zero-initialized.

It could be the case, but with the current `rcl_remap_t` structure and API implementation it doesn't add much value. The init/fini cycle would still be coupled to the parse API. However, if `rcl_remap_parse()` was capable of reusing `rcl_remap_t` structures (i.e. reuse past allocations), and if there was an `rcl_remap_move()` API to copy rules around cheaply, then relocating the init/fini cycle could save up quite a few unnecessary allocation/deallocation cycles. I didn't want to go that far.

Reply:
> We do, [right above](https://github.com/ros2/rcl/blob/6040298b429784ca9eb4d0404543d6211fd5f81d/rcl/src/rcl/arguments.c#L1792-L1800). Thing is, `remap_init` doesn't exist.

Oh, now I see.  OK, so that block of code is essentially `remap_init`, it just doesn't have that name.

> Right now it's not necessary. `_rcl_remap_parse_rule` takes a zero-initialized `rcl_remap_t` instance. On failure, it should stay zero-initialized.

I see.  So this PR is a cleanup of the allocated memory on failure further down.

I still find the way this is done weird.  The `rcl_remap_t` is currently an "inout" parameter, but it seems to me that it should just be an "out" parameter.  If we consider it more like that, then the caller can pass it in uninitialized.  This method would initialize it and add to it on success, and the caller would (eventually) need to free it.  On error, the parameter would still be considered "uninitialized".

The above just makes more sense to me.  That being said, I don't know how much you want to refactor this here.  So I'll leave it up to you whether you want to do that or just go with this code as-is.

Reply:
> I see. So this PR is a cleanup of the allocated memory on failure further down.

Exactly.

> I still find the way this is done weird. The rcl_remap_t is currently an "inout" parameter, but it seems to me that it should just be an "out" parameter. If we consider it more like that, then the caller can pass it in uninitialized. This method would initialize it and add to it on success, and the caller would (eventually) need to free it. On error, the parameter would still be considered "uninitialized".

That's almost exactly how it is handled right now, except that `_rcl_remap_parse_rule` does not explicitly zero-initialize the given `rcl_remap_t` structure, it assumes it is. So in a way it is an output parameter.